### PR TITLE
Read buffer to avoid byte-by-byte syscalls

### LINKED ISSUE #140: Reads HTTP response byte-by-byte, one recvfrom syscall per byte
I wrote the following simple test program for ureq:
```rust
fn main() -> anyhow::Result<()> {
    let response = ureq::get("http://httpbin.org/get").call();
    if response.ok() {
        let s = response.into_string()?;
        println!("Success: {}", s);
    } else {
        eprintln!("Error: {}", response.status_line());
    }
    Ok(())
}
```

Running that under strace shows the following:
```text
connect(3, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("35.170.21.246")}, 16) = 0
setsockopt(3, SOL_SOCKET, SO_RCVTIMEO_OLD, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
setsockopt(3, SOL_SOCKET, SO_SNDTIMEO_OLD, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
sendto(3, "GET /get HTTP/1.1\r\nHost: httpbin"..., 71, MSG_NOSIGNAL, NULL, 0) = 71
recvfrom(3, "H", 1, 0, NULL, NULL)      = 1
recvfrom(3, "T", 1, 0, NULL, NULL)      = 1
recvfrom(3, "T", 1, 0, NULL, NULL)      = 1
recvfrom(3, "P", 1, 0, NULL, NULL)      = 1
recvfrom(3, "/", 1, 0, NULL, NULL)      = 1
recvfrom(3, "1", 1, 0, NULL, NULL)      = 1
recvfrom(3, ".", 1, 0, NULL, NULL)      = 1
recvfrom(3, "1", 1, 0, NULL, NULL)      = 1
recvfrom(3, " ", 1, 0, NULL, NULL)      = 1
recvfrom(3, "2", 1, 0, NULL, NULL)      = 1
recvfrom(3, "0", 1, 0, NULL, NULL)      = 1
recvfrom(3, "0", 1, 0, NULL, NULL)      = 1
recvfrom(3, " ", 1, 0, NULL, NULL)      = 1
recvfrom(3, "O", 1, 0, NULL, NULL)      = 1
recvfrom(3, "K", 1, 0, NULL, NULL)      = 1
recvfrom(3, "\r", 1, 0, NULL, NULL)     = 1
recvfrom(3, "\n", 1, 0, NULL, NULL)     = 1
recvfrom(3, "D", 1, 0, NULL, NULL)      = 1
recvfrom(3, "a", 1, 0, NULL, NULL)      = 1
recvfrom(3, "t", 1, 0, NULL, NULL)      = 1
recvfrom(3, "e", 1, 0, NULL, NULL)      = 1
recvfrom(3, ":", 1, 0, NULL, NULL)      = 1
...
recvfrom(3, "A", 1, 0, NULL, NULL)      = 1
recvfrom(3, "c", 1, 0, NULL, NULL)      = 1
recvfrom(3, "c", 1, 0, NULL, NULL)      = 1
recvfrom(3, "e", 1, 0, NULL, NULL)      = 1
recvfrom(3, "s", 1, 0, NULL, NULL)      = 1
recvfrom(3, "s", 1, 0, NULL, NULL)      = 1
recvfrom(3, "-", 1, 0, NULL, NULL)      = 1
recvfrom(3, "C", 1, 0, NULL, NULL)      = 1
recvfrom(3, "o", 1, 0, NULL, NULL)      = 1
recvfrom(3, "n", 1, 0, NULL, NULL)      = 1
recvfrom(3, "t", 1, 0, NULL, NULL)      = 1
recvfrom(3, "r", 1, 0, NULL, NULL)      = 1
recvfrom(3, "o", 1, 0, NULL, NULL)      = 1
recvfrom(3, "l", 1, 0, NULL, NULL)      = 1
recvfrom(3, "-", 1, 0, NULL, NULL)      = 1
recvfrom(3, "A", 1, 0, NULL, NULL)      = 1
recvfrom(3, "l", 1, 0, NULL, NULL)      = 1
recvfrom(3, "l", 1, 0, NULL, NULL)      = 1
recvfrom(3, "o", 1, 0, NULL, NULL)      = 1
recvfrom(3, "w", 1, 0, NULL, NULL)      = 1
recvfrom(3, "-", 1, 0, NULL, NULL)      = 1
recvfrom(3, "C", 1, 0, NULL, NULL)      = 1
recvfrom(3, "r", 1, 0, NULL, NULL)      = 1
recvfrom(3, "e", 1, 0, NULL, NULL)      = 1
recvfrom(3, "d", 1, 0, NULL, NULL)      = 1
recvfrom(3, "e", 1, 0, NULL, NULL)      = 1
recvfrom(3, "n", 1, 0, NULL, NULL)      = 1
recvfrom(3, "t", 1, 0, NULL, NULL)      = 1
recvfrom(3, "i", 1, 0, NULL, NULL)      = 1
recvfrom(3, "a", 1, 0, NULL, NULL)      = 1
recvfrom(3, "l", 1, 0, NULL, NULL)      = 1
recvfrom(3, "s", 1, 0, NULL, NULL)      = 1
recvfrom(3, ":", 1, 0, NULL, NULL)      = 1
recvfrom(3, " ", 1, 0, NULL, NULL)      = 1
recvfrom(3, "t", 1, 0, NULL, NULL)      = 1
recvfrom(3, "r", 1, 0, NULL, NULL)      = 1
recvfrom(3, "u", 1, 0, NULL, NULL)      = 1
recvfrom(3, "e", 1, 0, NULL, NULL)      = 1
recvfrom(3, "\r", 1, 0, NULL, NULL)     = 1
recvfrom(3, "\n", 1, 0, NULL, NULL)     = 1
recvfrom(3, "\r", 1, 0, NULL, NULL)     = 1
recvfrom(3, "\n", 1, 0, NULL, NULL)     = 1
setsockopt(3, SOL_SOCKET, SO_RCVTIMEO_OLD, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
setsockopt(3, SOL_SOCKET, SO_SNDTIMEO_OLD, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
recvfrom(3, "{\n  \"args\": {}, \n  \"headers\": {\n", 32, 0, NULL, NULL) = 32
recvfrom(3, "    \"Accept\": \"*/*\", \n    \"Host\"", 32, 0, NULL, NULL) = 32
recvfrom(3, ": \"httpbin.org\", \n    \"User-Agen"..., 64, 0, NULL, NULL) = 64
recvfrom(3, "d\": \"Root=1-5f5576cf-9fa97f2fb31"..., 119, 0, NULL, NULL) = 119
setsockopt(3, SOL_SOCKET, SO_RCVTIMEO_OLD, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
setsockopt(3, SOL_SOCKET, SO_SNDTIMEO_OLD, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
close(3)                                = 0
```

ureq appears to be reading each byte of the headers from the socket with a separate `recvfrom` syscall. Once it gets through the headers, it uses slightly larger reads, but still very small buffers.

### PR DESCRIPTION
**Pull request by algesten**
Fixes #140

### CODE REVIEW

**Comment by algesten** [comment]
@jsha Does this look ok to you?

**Review by jsha — COMMENTED** [review]
Nice work! Thanks for jumping on a fix so fast.

Looking at this from a high level, there are a few things we need Stream or related structs to do:

 1. Allow us to read headers without having to go byte-by-byte in terms of underlying syscalls.
 2. Provide a Read implementation that can be passed to clients, which offers EOF when a response body has been fully read, and then hands ownership back to an Agent (LimitedReader / PoolReturnReader).
 3. Before each call to `.read()` on the underlying socket (that is, before each syscall), apply a timeout with `set_read_timeout`.

The buffering has to be applied at the lowest level: the object that gets returned to the Agent. That's because the buffer may have more bytes in it than are accounted for in a single response body, and we don't want to drop those on the floor. For instance, this could happen if a server sends Content-Length: 10, and then sends us 20 bytes of body. Those extra bytes need to cause a connection-closing error rather than being dropped on the floor. That's what you've done here, so great.

The response-limiting / pooling function (2) can be provided at a higher level.

The DeadlineStream is a little challenging. It wants to be at a high level, because the deadline is a property of a request, not a stream. But it also wants to be at a low level, because we'd like to call `set_read_timeout` only once per actual read syscall, not once per read against the BufReader. I think for now we can punt: We only apply wrap a stream in DeadlineStream before returning it to the user or running into_string etc, and in those cases the calling code is probably doing good-sized reads. We're mainly concerned here about our own header-reading code. With that said - is there a reason DeadlineStream needs to implement BufRead in this PR?

File: src/response.rs [hunk 1]
```diff
@@ -590,17 +590,25 @@ pub(crate) fn set_stream(resp: &mut Response, url: String, unit: Option<Unit>, s
     resp.stream = Some(stream);
 }
 
-fn read_next_line<R: Read>(reader: &mut R) -> IoResult<String> {
+fn read_next_line<R: BufRead>(reader: &mut R) -> IoResult<String> {
```
[the remark this answers is not in the document]
Comment (algesten) [reply src/response.rs:593]: Good point! I fix!

File: src/response.rs [hunk 2]
```diff
@@ -469,11 +469,11 @@ impl Response {
     ///
     /// assert_eq!(resp.status(), 401);
     /// ```
-    pub fn from_read(reader: impl Read) -> Self {
+    pub fn from_read(reader: impl BufRead) -> Self {
```
[the remark this answers is not in the document]
Comment (algesten) [reply src/response.rs:472]: 👍

File: src/stream.rs [hunk 3]
```diff
@@ -25,14 +27,14 @@ use crate::unit::Unit;
 
 #[allow(clippy::large_enum_variant)]
 pub enum Stream {
```
[the remark this answers is not in the document]
Comment (algesten) [reply src/stream.rs:29]: It's not.

ureq was made before `pub(crate)` was a thing in rust. `Stream` was never public API, but public internally in ureq (as opposed to file private). There are a number of places where can tidy this up.

File: src/stream.rs [hunk 4]
```diff
@@ -86,6 +88,16 @@ impl Read for DeadlineStream {
     }
 }
 
+impl BufRead for DeadlineStream {
+    fn fill_buf(&mut self) -> IoResult<&[u8]> {
+        self.stream.fill_buf()
```
[the remark this answers is not in the document]
Comment (algesten) [reply src/stream.rs:93]: ok!

**Comment by algesten** [comment]
@jsha we have a contradiction in requirements here.

`pub fn from_read(reader: impl BufRead) -> Self {`

> Since this is a public function, I think it would be better to continue to accept Read, and wrap it with a BufReader ourselves.

I'm not sure that would work. If we wrap an `impl Read` in `BufReader`, it's likely we over-read/buffer while parsing the header and then just drop the bytes in the buffer.

But you're right, this is public API, so it's a breaking change, which isn't good. One option would be to retain the old byte-by-byte thing for backwards compatibility of `Response::from_read()`, but then make a new `Response::from_buf_read()` that we use internally.

------

```rust
    let mut stream = stream::DeadlineStream::new(stream, unit.deadline);
    let mut resp = Response::from_read(&mut stream);
   …
    crate::response::set_stream(&mut resp, unit.url.to_string(), Some(unit), stream.into());
```

This is where the `DeadlineStream`. We can't really wrap DeadlineStream in a BufReader, cause, as you pointed out, the buffering must be done as far down as possible or we'll drop bytes.

I wonder whether we should try and merge the function of `DeadlineStream` into `Stream` to avoid this wrapping and unwrapping?

Regardless of whether we do that, `fill_buf` could maybe also propagate the timeouts in case it will result in reads?

**Comment by algesten** [comment]
> One option would be to retain the old byte-by-byte thing for backwards compatibility of Response::from_read(), but then make a new Response::from_buf_read() that we use internally.

Actually, this might not be necessary. `Read::read` on a `BufReader` uses `fill_buf` internally, so even we just retain the `Read` trait (and keep the one-by-one function), we will still avoid having individual syscalls for every byte.

**Comment by algesten** [comment]
@jsha I think my commit fixed all of it.

1. Remove `BufRead` from `DeadlineStream`.
2. Return to `Read` for `Response::from_read`

It means we're not breaking API but actually are buffering because `read` uses `fill_buf` inside `BufReader`.

The only drawback might be that we're setting the timeouts on the socket also when there will be no reads (since byte-by-byte will mostly hit the internal buffer, but each byte will set the timeouts).

**Review by jsha — APPROVED** [review]
Nice, thanks for making those changes. This looks good, and I tested it locally and it works as expected. It's a little awkward that we represent a read/write stream with a BufReader, but getting the underlying stream with get_mut seems to work fine, so no big deal.

**Pull request merged**